In [3]:
!pip install scikit-learn
!pip install matplotlib

In [7]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score

digits = load_digits()
digits_data = digits.data
digits_label = digits.target

# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_label,
                                                    test_size=0.2,
                                                    random_state=7)

# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("decision tree:\n", classification_report(y_test, y_pred))
print("accuracy : ", accuracy, "\n")

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("random_forest:\n ", classification_report(y_test, y_pred))
print("accuracy : ", accuracy, "\n")

SupportVectorMachine = svm.SVC(kernel='linear')
SupportVectorMachine.fit(X_train, y_train)
y_pred = SupportVectorMachine.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("SupportVectorMachine: \n", classification_report(y_test, y_pred))
print("accuracy : ", accuracy, "\n")

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("sgd_model: \n", classification_report(y_test, y_pred))
print("accuracy : ", accuracy, "\n")


decision tree:
               precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360

accuracy :  0.8555555555555555 

random_forest:
                precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2


가장 성능이 좋은 분류기는 0.96의 정확도를 낸 random_forest라고 생각한다.
이진 분류가 아닌이상 오차행렬로 정확도를 판단하긴 어렵다 생각해서, 
accuracy가 높으면 가장 좋은 분류기라고 생각했기 때문이다.
